File: engine/swiftui/Sources/MilanoSDK/Rendering/MilanoNode.swift
import SwiftUI

/// What a renderer receives: the node's resolved values, its materialized
/// children, and the emission surface.
public struct MilanoNode {
    let core: MilanoViewCore
    let resolved: ResolvedNode

    /// The component type name.
    public var type: String { resolved.type }

    /// The node's id, or canonical path when no id is declared.
    public var reference: String { resolved.reference }

    /// The resolved value of a declared property. Because the gate
    /// type-checked everything, reading with the declared type's accessor
    /// always succeeds; an absent optional is `.null`.
    public func property(_ name: String) -> MilanoValue {
        resolved.values[name] ?? .null
    }

    /// The node's materialized children, ready to place. Each is
    /// Identifiable by its node reference.
    public var children: [MilanoChildView] {
        resolved.children.map { MilanoChildView(core: core, resolved: $0) }
    }

    /// Emits a declared event into dispatch; invalid emissions are dropped
    /// and reported per Foundations.
    public func emit(_ event: String, payload: MilanoValue? = nil) {
        core.emit(node: resolved.reference, event: event, payload: payload)
    }

    /// Reports a widget interaction to the engine's user-interaction
    /// stream, for signals the document does not model as events (focus,
    /// visibility, selection). Never touches dispatch or state; a no-op
    /// when the engine carries no user-interaction observer.
    public func userInteraction(
        _ kind: MilanoUserInteraction.Kind, value: MilanoValue? = nil
    ) {
        core.record(kind, node: resolved.reference, name: nil, value: value)
    }
}

/// One materialized child, ready to place in a layout.
public struct MilanoChildView: View, Identifiable {
    nonisolated let core: MilanoViewCore
    nonisolated let resolved: ResolvedNode

    public nonisolated var id: String { resolved.reference }

    public var body: some View {
        milanoRender(core: core, resolved: resolved)
    }
}

File: engine/compose/src/commonMain/kotlin/dev/getmilano/rendering/MilanoNode.kt
package dev.getmilano

import androidx.compose.runtime.Composable

/**
 * What a renderer receives: the node's resolved values, its materialized
 * children, and the emission surface.
 */
class MilanoNode internal constructor(
    private val view: MilanoView,
    private val resolved: ResolvedNode,
) {
    /** The component type name. */
    val type: String get() = resolved.type

    /** The node's id, or canonical path when no id is declared. */
    val reference: String get() = resolved.reference

    /**
     * The resolved value of a declared property. Because the gate
     * type-checked everything, reading with the declared type's accessor
     * always succeeds; an absent optional is Null.
     */
    fun property(name: String): MilanoValue = resolved.values[name] ?: MilanoValue.Null

    /**
     * The node's materialized children, ready to place. Each carries a
     * stable key: its node reference.
     */
    val children: List<MilanoChild>
        get() = resolved.children.map { MilanoChild(view, it) }

    /**
     * Emits a declared event into dispatch; invalid emissions are dropped
     * and reported per Foundations.
     */
    fun emit(
        event: String,
        payload: MilanoValue? = null,
    ) {
        view.emit(resolved.reference, event, payload)
    }

    /**
     * Reports a widget interaction to the engine's user-interaction
     * stream, for signals the document does not model as events (focus,
     * visibility, selection). Never touches dispatch or state; a no-op
     * when the engine carries no user-interaction observer.
     */
    fun userInteraction(
        kind: MilanoUserInteraction.Kind,
        value: MilanoValue? = null,
    ) {
        view.record(kind, resolved.reference, null, value)
    }
}

/** One materialized child, ready to place in a layout. */
class MilanoChild internal constructor(
    private val view: MilanoView,
    private val resolved: ResolvedNode,
) {
    /** Stable identity for keyed composition: the node reference. */
    val key: String get() = resolved.reference

    @Composable
    fun Render() {
        RenderNode(view, resolved)
    }
}

/**
 * Dispatches one resolved node to its registered renderer (or the
 * placeholder renderer). Registry coverage is total by construction, so
 * the lookups cannot miss.
 */
@Composable
internal fun RenderNode(
    view: MilanoView,
    resolved: ResolvedNode,
) {
    if (resolved.isPlaceholder) {
        view.engine.registry.placeholder?.Render(
            MilanoUnknownNode(
                type = resolved.type,
                reference = resolved.reference,
                rawSubtree = resolved.rawSubtree ?: MilanoValue.Null,
            ),
        )
        return
    }
    view.engine.registry.renderers[resolved.type]
        ?.Render(MilanoNode(view, resolved))
}
